design-system/src/theme-provider.tsx

```diff
@@ -0,0 +1,94 @@
+import { useLayoutEffect, useState, useRef, useEffect } from 'react';
+import kebabCase from 'lodash/kebabCase';
+import isObject from 'lodash/isObject';
+import merge from 'lodash/merge';
+import { themes, themesNames } from './custom-properties';
+
+const allThemesNames = Object.keys(themesNames);
+
+type ThemeName = keyof typeof themes;
+
+const toVars = (obj: Record<string, string>) =>
+  Object.fromEntries(
+    Object.entries(obj).map(([key, value]) => [`--${kebabCase(key)}`, value])
+  );
+
+// used to cover SSR builds (for instance in Gatsby)
+const isBrowser = typeof window !== 'undefined';
+
+const defaultParentSelector = (): HTMLElement | null =>
+  document.querySelector(':root');
+
+type TApplyTheme = {
+  newTheme?: string;
+  parentSelector: typeof defaultParentSelector;
+  themeOverrides?: Record<string, string>;
+};
+
+const applyTheme = ({
+  newTheme,
+  parentSelector = defaultParentSelector,
+  themeOverrides,
+}: TApplyTheme): void => {
+  const target = isBrowser ? parentSelector() : null;
+
+  // With no target we can't change themes
+  if (!target) return;
+
+  const validTheme = (
+    allThemesNames.includes(newTheme || '') ? newTheme! : 'default'
+  ) as ThemeName;
+  if (newTheme && newTheme !== validTheme) {
+    console.warn(
+      `ThemeProvider: the specified theme '${newTheme}' is not supported.`
+    );
+  }
+
+  const vars = toVars(
+    themeOverrides && isObject(themeOverrides)
+      ? merge({}, themes.default, themes[validTheme], themeOverrides)
+      : themes[validTheme]
+  );
+
+  Object.entries(vars).forEach(([key, value]) => {
+    target.style.setProperty(key, value);
+  });
+  target.setAttribute('data-theme', validTheme);
+};
+
+type ThemeProviderProps = {
+  parentSelector: typeof defaultParentSelector;
+  theme?: string;
+  themeOverrides?: Record<string, string>;
+};
+
+const ThemeProvider = (props: ThemeProviderProps) => {
+  const parentSelectorRef = useRef(props.parentSelector);
+  useLayoutEffect(() => {
+    applyTheme({
+      newTheme: props.theme,
+      parentSelector: parentSelectorRef.current,
+      themeOverrides: props.themeOverrides,
+    });
+  }, [props.theme, props.themeOverrides]);
```

Comment:
I was wondering a bit about this, as `props.themeOverrides` could be passed as inline object in the props, causing the reference to always change (thus executing the effect on each render).

Can you think of an approach we can use to avoid that?

Reply:
I've opened [this PR](https://github.com/commercetools/ui-kit/pull/2272) with a proposal.